csa.rs

```diff
@@ -0,0 +1,134 @@
+use std::io;
+
+const MAX_STATIONS: usize = 100000;
+
+#[derive(Debug)]
+struct Connection {
+    departure_station: usize,
+    arrival_station: usize,
+    departure_timestamp: u32,
+    arrival_timestamp: u32
+}
+
+impl Connection {
+    fn parse(line: String) -> Connection {
+        let splitted = line.split(" ")
+            .map(|bite| {
+                bite.parse()
+                    .ok()
+                    .expect(&format!("failed to read {} as integer", bite))
+            })
+            .collect::<Vec<u32>>();
+        
+        Connection {
+            departure_station: splitted[0] as usize,
+            arrival_station: splitted[1] as usize,
+            departure_timestamp: splitted[2],
+            arrival_timestamp: splitted[3]
+        }
+    }
+}
+
+fn csa_main_loop(timetable: &Vec<Connection>, arrival_station: usize, earliest_arrival: &mut [u32], in_connection: &mut [usize]) {
+    let mut earliest = std::u32::MAX;
+        
+    for (i, connection) in timetable.iter().enumerate() {
+        if connection.departure_timestamp >= earliest_arrival[connection.departure_station] &&
+                connection.arrival_timestamp < earliest_arrival[connection.arrival_station] {
```

Comment:
Each time you do a `v[i]`, you do a bound check, so better to do that once (even if the compiler may be clever).